tests/snuba/search/test_backend.py

```diff
@@ -1217,6 +1229,354 @@ def test_first_release_environments(self):
+
+        results = self.make_query(
+            environments=[prod_env],
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([group_c])
+
+        results = self.make_query(
+            environments=[staging_env, prod_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            environments=[staging_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            environments=[prod_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([])
+
+        #  groupenv(group_a, prod, release_2)
+
+        group_a_prod_env = GroupEnvironment.get_or_create(
+            group_id=group_a.id,
+            environment_id=prod_env.id,
+        )[0]
+
+        group_a_prod_env.first_release = release_2
+        group_a_prod_env.save()
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([group_a, group_b])
+
+        results = self.make_query(
+            environments=[staging_env, prod_env],
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        # note that group_b does not show up since it has no events that occured in any environments
+        assert set(results) == set([group_a])
+
+        results = self.make_query(
+            environments=[staging_env],
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([group_a])
+
+        results = self.make_query(
+            environments=[prod_env],
+            search_filter_query='first_release:%s' % release_1.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([group_a, group_c])
+
+        results = self.make_query(
+            environments=[staging_env, prod_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([group_a])
+
+        results = self.make_query(
+            environments=[staging_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([])
+
+        results = self.make_query(
+            environments=[prod_env],
+            search_filter_query='first_release:%s' % release_2.version,
+        )
+        assert set(results) == set([group_a])
```

Comment:
I think the assertions build up could be shorter/simpler if you built all the state at once based on what you outline in your comment and then do all the queries to ensure those results are correct. The various intermediary states are less valuable when you repeat many of them as more state is added.

Reply:
Yea I'm going to  consider this approach on the next commit.